src/analysis/normed/group/add_circle.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.order.floor
+import algebra.order.to_interval_mod
+import analysis.normed.group.quotient
+
+/-!
+# The additive circle
+
+We define the additive circle `add_circle p` as the quotient `𝕜 ⧸ (ℤ ∙ p)` for some period `p : 𝕜`.
+Note that when `𝕜 = ℝ`, this is naturally a normed group; for example if `p = 1` then:
+`∥(x : add_circle 1)∥ = |x - round x|` for any `x : ℝ` (see `unit_add_circle.norm_eq`).
+
+See also `circle` and `real.angle`.
+
+## Main definitions:
+
+ * `add_circle`: the additive circle `𝕜 ⧸ (ℤ ∙ p)` for some period `p : 𝕜`
+ * `unit_add_circle`: the special case `ℝ ⧸ ℤ`
+ * `add_circle.equiv_add_circle`: the rescaling equivalence `add_circle p ≃+ add_circle q`
+ * `add_circle.equiv_Ico`: the natural equivalence `add_circle p ≃ Ico 0 p`
+ * `add_circle.norm_eq`: a characterisation of the norm on `add_circle p`
+
+## Implementation notes:
+
+Although the most important case is `𝕜 = ℝ` we wish to support other types of scalars, such as
+the rational circle `add_circle (1 : ℚ)`, and so we set things up more generally.
+
+## TODO
+
+ * Link with periodicity
+ * Measure space structure
+ * Lie group structure
+ * Exponential equivalence to `circle`
+ * The fact `inner_product_geometry.angle (real.cos θ) (real.sin θ) = ∥(θ : real.angle)∥`
+
+-/
+
+noncomputable theory
+
+open set int (hiding mem_zmultiples_iff) add_subgroup
+
+variables {𝕜 : Type*}
+
+/-- The "additive circle": `𝕜 ⧸ (ℤ ∙ p)`. See also `circle` and `real.angle`. -/
+@[derive [add_comm_group, topological_space, topological_add_group, inhabited, has_coe_t 𝕜],
+  nolint unused_arguments]
+def add_circle [linear_ordered_add_comm_group 𝕜] [topological_space 𝕜] [order_topology 𝕜] (p : 𝕜) :=
+𝕜 ⧸ zmultiples p
+
+namespace add_circle
+
+section linear_ordered_field
+
+variables [linear_ordered_field 𝕜] [topological_space 𝕜] [order_topology 𝕜] (p q : 𝕜)
+
+@[continuity, nolint unused_arguments] protected lemma continuous_mk' :
+  continuous (quotient_add_group.mk' (zmultiples p) : 𝕜 → add_circle p) :=
+continuous_coinduced_rng
+
+/-- An auxiliary definition used only for constructing `add_circle.equiv_add_circle`. -/
+private def equiv_add_circle_aux (hp : p ≠ 0) : add_circle p →+ add_circle q :=
+quotient_add_group.lift _
```

Comment:
It can certainly be used but at least my approach looks less promising than what is here:
```lean
have h : function.periodic (λ (x : 𝕜), (↑(q * (p⁻¹ * x)) : add_circle q)) p := λ x, by simp [mul_add, hp],
{ to_fun := h.lift,
  map_zero' := by { erw h.lift_coe 0, simp, },
  map_add' := sorry, }
```